arrow/src/main/java/org/apache/iceberg/arrow/DictEncodedArrowConverter.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.iceberg.arrow.vectorized.ArrowVectorAccessor;
+import org.apache.iceberg.arrow.vectorized.VectorHolder;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/**
+ * This converts dictionary encoded arrow vectors to a correctly typed arrow vector.
+ */
+public class DictEncodedArrowConverter {
+
+  private DictEncodedArrowConverter() {
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static FieldVector toArrowVector(VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Preconditions.checkArgument(null != vectorHolder, "VectorHolder cannot be null");
+    Preconditions.checkArgument(null != accessor, "ArrowVectorAccessor cannot be null");
+    if (vectorHolder.isDictionaryEncoded()) {
+      if (Type.TypeID.DECIMAL.equals(vectorHolder.icebergType().typeId())) {
+        return toDecimalVector(vectorHolder, accessor);
+      } else if (Type.TypeID.TIMESTAMP.equals(vectorHolder.icebergType().typeId())) {
+        return toTimestampVector(vectorHolder, accessor);
+      } else if (Type.TypeID.LONG.equals(vectorHolder.icebergType().typeId())) {
+        return toBigIntVector(vectorHolder, accessor);
+      } else if (Type.TypeID.FLOAT.equals(vectorHolder.icebergType().typeId())) {
+        return toFloat4Vector(vectorHolder, accessor);
+      } else if (Type.TypeID.DOUBLE.equals(vectorHolder.icebergType().typeId())) {
+        return toFloat8Vector(vectorHolder, accessor);
+      } else if (Type.TypeID.STRING.equals(vectorHolder.icebergType().typeId())) {
+        return toVarCharVector(vectorHolder, accessor);
+      } else if (Type.TypeID.BINARY.equals(vectorHolder.icebergType().typeId())) {
+        return toVarBinaryVector(vectorHolder, accessor);
+      } else if (Type.TypeID.TIME.equals(vectorHolder.icebergType().typeId())) {
+        return toTimeMicroVector(vectorHolder, accessor);
+      }
+
+      throw new IllegalArgumentException(String.format("Cannot convert dict encoded field '%s' of type '%s' to Arrow " +
+              "vector as it is currently not supported",
+          vectorHolder.icebergField().name(), vectorHolder.icebergType().typeId()));
+    }
+
+    return vectorHolder.vector();
+  }
+
+  private static DecimalVector toDecimalVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    int precision = ((Types.DecimalType) vectorHolder.icebergType()).precision();
+    int scale = ((Types.DecimalType) vectorHolder.icebergType()).scale();
+
+    DecimalVector vector = new DecimalVector(
+        vectorHolder.vector().getName(),
+        ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+        vectorHolder.vector().getAllocator());
+
+    for (int i = 0; i < vectorHolder.vector().getValueCount(); i++) {
+      vector.setSafe(i, (BigDecimal) accessor.getDecimal(i, precision, scale));
```

Comment:
since we already know the vector size, can we call allocateNew(size) and then use vector.set() instead of setSafe()? This will give avoid unnecessary safety checks and should give better performance.

Reply:
done

Reply:
thank you